drivers/espi/espi_realtek_rts5912.c

```diff
@@ -1242,6 +1468,92 @@ static void espi_vw_ch_setup(const struct device *dev)
 	irq_enable(DT_IRQ_BY_NAME(DT_DRV_INST(0), vw_idx61, irq));
 }
 
+#ifdef CONFIG_ESPI_PERIPHERAL_8042_KBC
+
+#define ESPI_VW_EVENT_RETRY_CNT   1024UL
+#define ESPI_VW_EVENT_TIMEOUT_CNT 10000UL
+
+static int espi_send_vw_event(uint8_t index, uint8_t data, const struct device *dev)
+{
+	const struct espi_rts5912_config *const espi_config = dev->config;
+	uint32_t i;
+	uint16_t retry_counter = ESPI_VW_EVENT_RETRY_CNT;
+	uint32_t timeout_cnt = 0;
+
+	volatile struct espi_reg *const espi_reg = espi_config->espi_reg;
+
+	if ((espi_reg->EVCFG & ESPI_EVCFG_CHEN) == 0 || (espi_reg->EVCFG & ESPI_EVCFG_CHRDY) == 0) {
+		return -EBUSY;
+	}
+
+	while (espi_reg->EVSTS & ESPI_EVSTS_TXFULL) {
+		timeout_cnt++;
+		if (timeout_cnt >= ESPI_VW_EVENT_TIMEOUT_CNT) {
+			return -EBUSY;
+		}
+	}
+
+	i = 0x0000FFFF & ((uint32_t)index << 8 | (uint32_t)data);
+	espi_reg->EVTXDAT = i;
+
+	while (espi_reg->EVSTS & ESPI_EVSTS_TXFULL && retry_counter) {
+		retry_counter--;
+	}
+
+	espi_reg->EVSTS |= ESPI_EVSTS_TXDONE;
+
+	if (retry_counter == 0) {
+		return -EBUSY;
+	}
+
+	return 0;
+}
+
+static void espi_send_vw_event_with_kbdata(uint8_t index, uint8_t data, uint32_t kbc_data,
+					   const struct device *dev)
+{
+	const struct espi_rts5912_config *const espi_config = dev->config;
+	struct espi_rts5912_data *const espi_data = dev->data;
+
+	uint32_t i;
+	uint32_t timeout_cnt = 0;
+
+	volatile struct espi_reg *const espi_reg = espi_config->espi_reg;
+	volatile struct kbc_reg *const kbc_reg = espi_config->kbc_reg;
+
+	if ((espi_reg->EVCFG & ESPI_EVCFG_CHEN) == 0 || (espi_reg->EVCFG & ESPI_EVCFG_CHRDY) == 0) {
+		return;
+	}
+
+	while (espi_reg->EVSTS & ESPI_EVSTS_TXFULL) {
+		timeout_cnt++;
+		if (timeout_cnt >= ESPI_VW_EVENT_TIMEOUT_CNT) {
+			return;
+		}
+	}
```

Comment:
Non-blocking: another place to use `WAIT_FOR()`

You also check the same condition `TXFULL` with the same timeout in multiple places.  This can be moved to a helper function.